tensorflow/contrib/go/tensor.go

```diff
@@ -0,0 +1,753 @@
+		dataType, _ = getDataTypeFromReflect(v.Type().Elem().Kind(), 1)
+		if dataType == DTString {
+			strings := make([]string, v.Len())
+			for i := 0; i < v.Len(); i++ {
+				strings[i] = v.Index(i).String()
+			}
+			buf := encodeStrings(strings)
+			return newTensor(DTString, TensorShape{{int64(len(strings))}},
+				unsafe.Pointer(&(buf[0])), int64(len(buf)))
+		}
+
+		dataSer, dims, dataType, dataSize, err = serialize(data, 0, [][]int64{})
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// Scalar tensor
+		dataSer = []interface{}{data}
+		dims = [][]int64{}
+		dataSize = int64(v.Type().Size())
+		if dataType, err = getDataTypeFromReflect(v.Kind(), dataSize); err != nil {
+			return nil, err
+		}
+	}
+	ts := TensorShape(dims)
+
+	auxTensor := new(Tensor)
+	switch dataType {
+	case DTFloat:
+		auxTensor.FloatVal = make([]float32, len(dataSer))
+		for i, v := range dataSer {
+			auxTensor.FloatVal[i] = v.(float32)
+		}
+		dataPtr = reflect.ValueOf(auxTensor.FloatVal).Pointer()
+	case DTDouble:
+		auxTensor.DoubleVal = make([]float64, len(dataSer))
+		for i, v := range dataSer {
+			auxTensor.DoubleVal[i] = v.(float64)
+		}
+		dataPtr = reflect.ValueOf(auxTensor.DoubleVal).Pointer()
+	case DTInt8, DTInt16, DTInt32, DTUint8:
+		auxTensor.IntVal = make([]int32, len(dataSer))
+		for i, v := range dataSer {
+			auxTensor.IntVal[i] = int32(reflect.ValueOf(v).Int())
+		}
+		dataPtr = reflect.ValueOf(auxTensor.IntVal).Pointer()
+	case DTInt64:
+		auxTensor.Int64Val = make([]int64, len(dataSer))
+		for i, v := range dataSer {
+			auxTensor.Int64Val[i] = reflect.ValueOf(v).Int()
+		}
+		dataPtr = reflect.ValueOf(auxTensor.Int64Val).Pointer()
+	case DTBool:
+		auxTensor.BoolVal = make([]bool, len(dataSer))
+		for i, v := range dataSer {
+			auxTensor.BoolVal[i] = v.(bool)
+		}
+		dataPtr = reflect.ValueOf(auxTensor.BoolVal).Pointer()
+	case DTString:
+		auxTensor.StringVal = make([][]byte, len(dataSer))
+		for i, v := range dataSer {
+			auxTensor.StringVal[i] = []byte(v.(string))
+		}
+		dataPtr = reflect.ValueOf(auxTensor.StringVal).Pointer()
+	default:
+		return nil, &ErrTensorTypeNotSupported{
+			TensotType: dataType,
+		}
+	}
+
+	tensor, err := newTensor(dataType, ts, unsafe.Pointer(dataPtr), int64(len(dataSer))*dataSize)
+	if err != nil {
+		return nil, err
+	}
+
+	tensor.FloatVal = auxTensor.FloatVal
+	tensor.DoubleVal = auxTensor.DoubleVal
+	tensor.IntVal = auxTensor.IntVal
+	tensor.StringVal = auxTensor.StringVal
+	tensor.ScomplexVal = auxTensor.ScomplexVal
+	tensor.Int64Val = auxTensor.Int64Val
+	tensor.BoolVal = auxTensor.BoolVal
+
+	return tensor, nil
+}
+
+// DataType returns the data type of the elements contained in the tensor.
+func (t *Tensor) DataType() DataType {
+	return DataType(TF_TensorType(t.tensor))
+}
+
+// NumDims returns the number of dimensions in tensor t.
+func (t *Tensor) NumDims() int {
+	return TF_NumDims(t.tensor)
+}
+
+// Shape returns the shape of the tensor.
+func (t *Tensor) Shape() TensorShape {
```

Comment:
In response to the earlier lldims question:  I'm now very confused about the representation of shapes in the API.  A Shape, in general, is just a slice of int64s that show the size in each dimension.  It should have NumDims() entries.  Why is it being represented as a [][]int64 here?

(Perhaps this is the source of confusion for calling newtensor?  lldims should just be an array of int64s of length num_dims.)

In my mind, this code should be:

shape := make([]int64, t.NumDims())
for i := 0; i < t.NumDims(); i++ {
  shape[i] = t.Dim(i)
}

Also, it's incorrect for t.Dim() to be of type int -- it *must* be type int64;  tensors may have more than max int32 entries in a single dimension.

Does this help clean up some of the confusion in the new tensor lldims path?